cmd/soroban-rpc/lib/preflight/src/fees.rs

```diff
@@ -292,118 +285,193 @@ fn storage_footprint_to_ledger_footprint(foot: &Footprint) -> Result<LedgerFootp
+        }
     };
-    Ok((transaction_data, min_fee))
+    Ok(res)
 }
 
 pub(crate) fn compute_restore_footprint_transaction_data_and_min_fee(
     footprint: LedgerFootprint,
-    snapshot_source: &ledger_storage::LedgerStorage,
+    ledger_storage: &ledger_storage::LedgerStorage,
     bucket_list_size: u64,
+    current_ledger_seq: u32,
 ) -> Result<(SorobanTransactionData, i64), Box<dyn error::Error>> {
+    let ConfigSettingEntry::StateExpiration(state_expiration) =
+        ledger_storage.get_configuration_setting(ConfigSettingId::StateExpiration)?
+        else {
+            return Err(
+                "get_fee_configuration(): unexpected config setting entry for StateExpiration key".into(),
+            );
+        };
+    // FIXME: we are querying the ledger entries from the DB twice (once here and again in calculate_unmodified_ledger_entry_bytes)
+    let mut rent_changes: Vec<LedgerEntryRentChange> = Vec::new();
+    for key in footprint.read_write.as_vec() {
+        let unmodified_entry = ledger_storage.get(key, true)?;
+        let (expiration_ledger, durability) =
+            get_bumpable_entry_expiration_and_durability(&unmodified_entry)?;
+        if durability != Persistent {
+            let err = format!("Non-persistent key ({:?}) in footprint", key).into();
+            return Err(err);
+        }
+        if current_ledger_seq < expiration_ledger {
+            // TODO: is this accurate?
+            //       shouldn't we just check that cd.expiration_ledger_seq < current_ledger_seq + min_persistent_expiration - 1 ?
+            // noop (the entry hadn't expired)
+            continue;
+        }
+        let new_expiration_ledger =
+            current_ledger_seq + state_expiration.min_persistent_entry_expiration - 1;
+        let size = (key.to_xdr()?.len() + unmodified_entry.to_xdr()?.len()) as u32;
+        let rent_change = LedgerEntryRentChange {
+            is_persistent: true,
+            // TODO: is this correct? Or should the old size be considered to be `size`?
+            old_size_bytes: 0,
+            new_size_bytes: size,
+            old_expiration_ledger: expiration_ledger,
+            new_expiration_ledger,
+        };
+        rent_changes.push(rent_change);
+    }
     let write_bytes = calculate_unmodified_ledger_entry_bytes(
         footprint.read_write.as_vec(),
-        snapshot_source,
+        ledger_storage,
         true,
     )?;
-    let soroban_resources = hack_soroban_resources(SorobanResources {
+    let soroban_resources = SorobanResources {
         footprint,
         instructions: 0,
-        // FIXME(fons): this seems to be a workaround a bug in code (the fix is to also count bytes read but not written in readBytes).
+        // FIXME(fons): this seems to be a workaround for a bug in Core (the fix is to also count bytes read but not written in readBytes).
         //        we should review it in preview 11.
         read_bytes: write_bytes,
         write_bytes,
```

Comment:
@sisuresh does this still hold?

Reply:
This is not a bug - the entries to restore are stored in ledger and thus need to be accounted in read bytes as well. The amounts are the same currently.